[LIBCLOUD-961] Update GCE public image projects list.

smarlowucf opened the pull request:
## Update GCE public image projects list

### Description

Add missing projects to list:
- cos-cloud
- suse-byos-cloud
- suse-sap-cloud
- windows-sql-cloud

Issue link: https://issues.apache.org/jira/browse/LIBCLOUD-961

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [ ] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [ ] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

smarlowucf committed:
[LIBCLOUD-961] Update GCE public image list.

Add missing projects and update the image families for each
project based on `gcloud compute images list` info.

pquentin commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1733,15 +1733,19 @@ class GCENodeDriver(NodeDriver):
         "opensuse-cloud": ["opensuse"],
         "rhel-cloud": ["rhel"],
         "suse-cloud": ["sles", "suse"],
+        "suse-byos-cloud": ["sles", "suse"],
```
The list may have been updated since you submitted the pull request, but I don't see suse-byos-cloud in https://cloud.google.com/compute/docs/images#os-compute-support. Do you think it's still needed?

smarlowucf replied:
Yes, it is still needed. It's not in the list however in the SUSE details below it's mentioned:

> If you need to reuse your existing SLES licenses to create Compute Engine instances, you can use Bring Your Own Subscription (BYOS) images available in the suse-byos-cloud project.

pquentin commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1733,15 +1733,19 @@ class GCENodeDriver(NodeDriver):
 
     IMAGE_PROJECTS = {
         "centos-cloud": ["centos"],
+        "cos-cloud": ["container-vm"],
```
Could we update the image families too? For example, in [the list you linked](https://cloud.google.com/compute/docs/images#os-compute-support), I see that the families associated with cos-cloud are cos-stable, cos-beta and cos-dev.

smarlowucf replied:
Yeah, I can update the image families.

smarlowucf replied:
On a side not: would it be possible to generate this list dynamically?

pquentin replied:
It can be interesting to generate the list dynamically if there's a JSON endpoint somewhere. I'm afraid crawling the HTML page is not worth the effort? I don't know.

pquentin reviewed (changes requested):
Thank you for your contribution! I have two questions.

smarlowucf commented:
@pquentin I've updated the list according gcloud tool. Removed the google containers project which is now cos-cloud and google nvme backports which doesn't appear in the project list anymore.

pquentin commented:
Thanks! This looks good, but the tests need to be updated: see https://travis-ci.org/apache/libcloud/builds/342458915. Please tell me if you need help here.

smarlowucf committed:
Update compute unit tests image families.

coreos should point to coreos-alpha, coreos-beta, or
coreos-stable.

smarlowucf commented:
@pquentin I started updating tests but ran into mocking errors:

`AttributeError: 'GCEMockHttp' object has no attribute '_global_images_family_coreos_stable'`

Skimming docs I don't see anything regarding the mocking solution and how to go about updating/recording new json responses.

pquentin commented:
@smarlowucf sorry this is something that we really need to document. See https://github.com/apache/libcloud/pull/1157#issuecomment-355883772 for an explanation in the context of another driver

pquentin commented:
Does that make sense?

smarlowucf commented:
> Does that make sense?

@pquentin Yep, thanks. I will take a look.

smarlowucf committed:
Update fixtures and re-add opensuse-cloud.

smarlowucf commented:
@pquentin The tests are passing but I'm not sure I did the fixtures correct. Also, this doesn't include any updated image lists for the new projects that were added.

Is there a way to record new json with image info for the new projects? `gcloud` doesn't seem to have a json option.

GwenaelPellenArkeup commented:
@smarlowucf  I may have misunderstood the question :

> Is there a way to record new json with image info for the new projects? gcloud doesn't seem to have a json option.

gcloud can export result in json format.

```bash
$ gcloud compute images list --format=json --filter='(family: "coreos")'

[
  {
    "archiveSizeBytes": "500540800",
    "creationTimestamp": "2018-03-02T18:02:37.081-08:00",
    "description": "CoreOS, CoreOS alpha, 1702.1.0, amd64-usr published on 2018-03-03",
    "diskSizeGb": "9",
    "family": "coreos-alpha",
    "guestOsFeatures": [
      {
        "type": "VIRTIO_SCSI_MULTIQUEUE"
      }
    ],
    "id": "4918486452405779667",
    "kind": "compute#image",
    "labelFingerprint": "42WmSpB8rSM=",
    "licenses": [
      "https://www.googleapis.com/compute/v1/projects/coreos-cloud/global/licenses/coreos-alpha"
    ],
    "name": "coreos-alpha-1702-1-0-v20180303",
    "rawDisk": {
      "containerType": "TAR",
      "source": ""
    },
    "selfLink": "https://www.googleapis.com/compute/v1/projects/coreos-cloud/global/images/coreos-alpha-1702-1-0-v20180303",
    "sourceType": "RAW",
    "status": "READY"
  },
  {
    "archiveSizeBytes": "490827904",
    "creationTimestamp": "2018-03-09T10:34:21.565-08:00",
    "description": "CoreOS, CoreOS beta, 1688.3.0, amd64-usr published on 2018-03-09",
    "diskSizeGb": "9",
    "family": "coreos-beta",
    "guestOsFeatures": [
      {
        "type": "VIRTIO_SCSI_MULTIQUEUE"
      }
    ],
    "id": "6827356011088538947",
    "kind": "compute#image",
    "labelFingerprint": "42WmSpB8rSM=",
    "licenses": [
      "https://www.googleapis.com/compute/v1/projects/coreos-cloud/global/licenses/coreos-beta"
    ],
    "name": "coreos-beta-1688-3-0-v20180309",
    "rawDisk": {
      "containerType": "TAR",
      "source": ""
    },
    "selfLink": "https://www.googleapis.com/compute/v1/projects/coreos-cloud/global/images/coreos-beta-1688-3-0-v20180309",
    "sourceType": "RAW",
    "status": "READY"
  },
  {
    "archiveSizeBytes": "470601088",
    "creationTimestamp": "2018-02-15T09:50:52.929-08:00",
    "description": "CoreOS, CoreOS stable, 1632.3.0, amd64-usr published on 2018-02-15",
    "diskSizeGb": "9",
    "family": "coreos-stable",
    "guestOsFeatures": [
      {
        "type": "VIRTIO_SCSI_MULTIQUEUE"
      }
    ],
    "id": "6217468005063358099",
    "kind": "compute#image",
    "labelFingerprint": "42WmSpB8rSM=",
    "licenses": [
      "https://www.googleapis.com/compute/v1/projects/coreos-cloud/global/licenses/coreos-stable"
    ],
    "name": "coreos-stable-1632-3-0-v20180215",
    "rawDisk": {
      "containerType": "TAR",
      "source": ""
    },
    "selfLink": "https://www.googleapis.com/compute/v1/projects/coreos-cloud/global/images/coreos-stable-1632-3-0-v20180215",
    "sourceType": "RAW",
    "status": "READY"
  }
]
```

smarlowucf commented:
> gcloud can export result in json format.

@GwenaelPellenArkeup Ah, thanks. I must have overlooked the format option... Will give it another shot.

smarlowucf committed:
Add json images list for new projects.

cos-cloud, suse-byos-cloud, suse-sap-cloud, windows-sql-cloud.

smarlowucf commented:
@pquentin I've added json image lists for the new projects. If this all looks good I will go ahead and merge/cleanup commit history.

pquentin commented:
Thanks! Merged in trunk. ✨ Really nice contribution.

I cleaned the history myself, I hope you don't mind. Can you please close LIBCLOUD-961?

smarlowucf commented:
> I cleaned the history myself, I hope you don't mind.

Yep, no worries. Thanks.

> Can you please close LIBCLOUD-961?

Closed.
